python/cudf/cudf/core/frame.py

```diff
@@ -1424,6 +1428,66 @@ def _apply_boolean_mask(self, boolean_mask):
         result._copy_type_metadata(self)
         return result
 
+    def _interpolate(
+        self,
+        method="linear",
+        axis=0,
+        limit=None,
+        inplace=False,
+        limit_direction=None,
+        limit_area=None,
+        downcast=None,
+        **kwargs,
+    ):
+        """
+        Interpolate data values between some points.
+
+        Parameters
+        ----------
+        method : str, default 'linear'
+            Interpolation technique to use. Currently,
+            only 'linear` is supported.
+            * 'linear': Ignore the index and treat the values as
+            equally spaced. This is the only method supported on MultiIndexes.
+            * 'index', 'values': linearly interpolate using the index as
+            an x-axis. Unsorted indices can lead to erroneous results.
+        axis : int, default 0
+            Axis to interpolate along. Currently,
+            only 'axis=0' is supprted.
+        inplace : bool, default False
+            Update the data in place if possible.
+
+        Returns
+        -------
+        Series or DataFrame
+            Returns the same object type as the caller, interpolated at
+            some or all ``NaN`` values
+
+        """
+
+        if method in {"pad", "ffill"} and limit_direction != "forward":
+            raise ValueError(
+                f"`limit_direction` must be 'forward' for method `{method}`"
+            )
+        if method in {"backfill", "bfill"} and limit_direction != "backward":
+            raise ValueError(
+                f"`limit_direction` must be 'backward' for method `{method}`"
+            )
+
+        columns = ColumnAccessor()
+
+        interpolator = cudf.core.algorithms.get_column_interpolator(method)
+        for colname, col in self._data.items():
+            if col.nullable:
+                col = col.fillna(np.nan)
+
+            # Interpolation methods may or may not need the index
+            to_interp = Frame(data={colname: col}, index=self.index)
+            result = interpolator(to_interp)
```

Comment:
It feels very awkward to force all interpolation functions to accept a `Frame` just so that they have the same API. Why not instead do:
```suggestion
            index = RangeIndex(start=0, stop=len(self), step=1) if method == 'linear' else self.index
            result = index_or_values_interpolation(col, index)
```
Then you can get rid of the `linear_interpolation` and `get_column_interpolator` functions. If we have more methods in the future we can consider a more appropriate dispatch mechanism at that point.

Reply:
I am really glad this came up and wasn't sure exactly what the right way forward was. There's a few reasons I tried to design this with "future state" in mind. There's a lot of [different](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.interpolate.html) `method`s in pandas. I imagine we *will* end up adding some in the future. Some of them aren't really interpolation in the precise sense (looking at `pad`) and are particular ways of simply filling NaNs based on the surrounding values. It seems like we could have a situation where some of our methods use `CuPy`, some use an actual libcudf `interpolate` api, and some wrap a string of `fill` operations that mimic the pandas result. In

I am not married to the interface though, really `index` and `values` are the odd ones out that need the index.

Reply:
Got it, definitely worth thinking about. Is it true that we can at least use a common API for all interpolators `def interpolate(column, index=None)`? If so, then I would at minimum prefer that interface to creating then immediately unpacking a `Frame` object, which was what first led me down this route. Then we can think about the question of how to dispatch to a particular method, although I would at minimum suggest to replace an `if`/`else` cascade with a dictionary lookup.